volume/dir.go

```diff
@@ -7,44 +7,36 @@ import (
 	"github.com/puppetlabs/wash/plugin"
 )
 
-// Dir represents a directory in a volume. It retains access to a map of directories
-// to their children and attribute data to populate subdirectories.
-//
-// TODO: by tying DirMap to the Node, it doesn't refresh until something triggers listing the resource
-// that produced the DirMap. Need to explore making this a callback that retrieves fresh data. How
-// should that interact with caching? We pass in DirMap so that we only have to retrieve it once
-// when constructing an entire hierarchy.
+// ListCB accepts a path and returns the content associated with that path.
+type ListCB = func(context.Context) (DirMap, error)
+
+// Dir represents a directory in a volume. It populates a subtree with listcb as needed.
 type Dir struct {
 	plugin.EntryBase
-	contentcb ContentCB
-	path      string
-	dirs      DirMap
+	cacheEntry plugin.Entry
+	listcb     ListCB
+	contentcb  ContentCB
+	path       string
 }
 
 // NewDir creates a Dir populated from dirs.
-func NewDir(name string, attr plugin.EntryAttributes, cb ContentCB, path string, dirs DirMap) *Dir {
+func NewDir(name string, attr plugin.EntryAttributes, cacheEntry plugin.Entry, lb ListCB, cb ContentCB, path string) *Dir {
 	vd := &Dir{
-		EntryBase: plugin.NewEntry(name),
-		contentcb: cb,
-		path:      path,
-		dirs:      dirs,
+		EntryBase:  plugin.NewEntry(name),
+		cacheEntry: cacheEntry,
```

Comment:
This package is getting a bit complicated with the callbacks + cacheEntry. It'd be nice to clean some of this up if possible, but that can be done in a separate PR.